File: src/main/java/com/api/twitter/security/application/usecases/UpdateUserCredentialsUseCase.java
package com.api.twitter.security.application.usecases;

import com.api.twitter.common.exception.BadRequestException;
import com.api.twitter.security.domain.model.UserCredentials;
import com.api.twitter.security.infrastructure.persistence.UserCredentialsRepository;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Component;

import java.util.UUID;

@Component
public class UpdateUserCredentialsUseCase {
    private PasswordEncoder passwordEncoder;
    private UserCredentialsRepository userCredentialsRepository;

    public UpdateUserCredentialsUseCase(PasswordEncoder passwordEncoder, UserCredentialsRepository userCredentialsRepository) {
        this.passwordEncoder = passwordEncoder;
        this.userCredentialsRepository = userCredentialsRepository;
    }

    public void execute(UUID userId, String username, String email) {
        UserCredentials userCredentials = userCredentialsRepository.findById(userId)
                .orElseThrow(() -> new BadRequestException("User not found"));
        userCredentials.setUsername(username);
        userCredentials.setEmail(email);
        userCredentials.validate();
        userCredentialsRepository.save(userCredentials);
    }

    public void updatePassword(UUID userId, String password){
        UserCredentials userCredentials = userCredentialsRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));

        userCredentials.setPassword(password);
        userCredentials.validatePassword();
        userCredentials.setPassword(passwordEncoder.encode(password));

        userCredentialsRepository.save(userCredentials);
    }
}

File: src/main/java/com/api/twitter/security/application/usecases/UpdateUserCredentialsUseCase.kt
package com.api.twitter.security.application.usecases

import com.api.twitter.common.exception.BadRequestException
import com.api.twitter.security.domain.model.UserCredentials
import com.api.twitter.security.infrastructure.persistence.UserCredentialsRepository
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Component
import java.util.*
import java.util.function.Supplier
import kotlin.jvm.optionals.getOrNull

@Component
class UpdateUserCredentialsUseCase(
    private val passwordEncoder: PasswordEncoder,
    private val userCredentialsRepository: UserCredentialsRepository
) {

    fun execute(userId: UUID, username: String, email: String) {
        val userCredentials = userCredentialsRepository.findById(userId).getOrNull()
            ?: throw BadRequestException("User not found")

        userCredentials.rawUsername = username
        userCredentials.email = email
        userCredentials.validate()
        userCredentialsRepository.save(userCredentials)
    }

    fun updatePassword(userId: UUID, password: String) {
        val userCredentials = userCredentialsRepository.findById(userId).getOrNull()
            ?: throw RuntimeException("User not found")

        userCredentials.rawPassword = password
        userCredentials.validatePassword()
        userCredentials.rawPassword = this.passwordEncoder.encode(password)
            ?: throw BadRequestException("Invalid password")

        userCredentialsRepository.save(userCredentials)
    }
}
